File: test/test_api_gfa1_lines.rb
require_relative "../lib/rgfa.rb"
require "test/unit"

TestAPI ||= Module.new

class TestAPI::GFA1Lines < Test::Unit::TestCase

  def test_C
    fields=["C","1","+","2","-","12","12M","MQ:i:1232","NM:i:3","ab:Z:abcd"]
    str=fields.join("\t")
    assert_nothing_raised { str.to_rgfa_line }
    assert_equal(RGFA::Line::Edge::Containment, str.to_rgfa_line.class)
    assert_equal(fields[0].to_sym, str.to_rgfa_line.record_type)
    assert_equal(fields[1].to_sym, str.to_rgfa_line.from)
    assert_equal(fields[2].to_sym, str.to_rgfa_line.from_orient)
    assert_equal(fields[3].to_sym, str.to_rgfa_line.to)
    assert_equal(fields[4].to_sym, str.to_rgfa_line.to_orient)
    assert_equal(12, str.to_rgfa_line.pos)
    assert_equal([RGFA::Alignment::CIGAR::Operation.new(12,:M)], str.to_rgfa_line.overlap)
    assert_equal(1232, str.to_rgfa_line.MQ)
    assert_equal(3, str.to_rgfa_line.NM)
    assert_equal("abcd", str.to_rgfa_line.ab)
    assert_raises(RGFA::FormatError) { (str+"\tH1").to_rgfa_line }
    assert_raises(RGFA::FormatError) { "C\tH".to_rgfa_line }
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[2]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[4]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[5]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[6]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::TypeError) do
      f=fields.dup; f[7]="MQ:Z:1232"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::TypeError) do
      f=fields.dup; f[8]="NM:Z:1232"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
  end

  def test_L
    fields=["L","1","+","2","-","12M","RC:i:1232","NM:i:3","ab:Z:abcd",
            "FC:i:2321","KC:i:1212","MQ:i:40"]
    str=fields.join("\t")
    assert_nothing_raised { str.to_rgfa_line }
    assert_equal(RGFA::Line::Edge::Link, str.to_rgfa_line.class)
    assert_equal(fields[0].to_sym, str.to_rgfa_line.record_type)
    assert_equal(fields[1].to_sym, str.to_rgfa_line.from)
    assert_equal(fields[2].to_sym, str.to_rgfa_line.from_orient)
    assert_equal(fields[3].to_sym, str.to_rgfa_line.to)
    assert_equal(fields[4].to_sym, str.to_rgfa_line.to_orient)
    assert_equal([RGFA::Alignment::CIGAR::Operation.new(12,:M)],
                 str.to_rgfa_line.overlap)
    assert_equal(1232, str.to_rgfa_line.RC)
    assert_equal(3, str.to_rgfa_line.NM)
    assert_equal(2321, str.to_rgfa_line.FC)
    assert_equal(1212, str.to_rgfa_line.KC)
    assert_equal(40, str.to_rgfa_line.MQ)
    assert_equal("abcd", str.to_rgfa_line.ab)
    assert_raises(RGFA::FormatError) { (str+"\tH1").to_rgfa_line }
    assert_raises(RGFA::FormatError) { "L\tH".to_rgfa_line }
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[2]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[4]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[5]="x"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::TypeError) do
      f=fields.dup; f[6]="RC:Z:1232"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::TypeError) do
      f=fields.dup; f[7]="NM:Z:1232"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
  end

  def test_L_coords
    g = RGFA.new(version: :gfa1)
    g << "S\t1\t*\tLN:i:100"
    g << "L\t1\t+\t2\t-\t1M2D10M1I"
    assert_equal([87,100], g.links[0].from_coords)
    assert_raises(RGFA::ValueError) {g.links[0].to_coords}
    g << "S\t2\t*\tLN:i:100"
    assert_equal([88,100], g.links[0].to_coords)
    g << "L\t3\t-\t4\t+\t10M2P3D1M"
    assert_equal([0,14], g.links[1].from_coords)
    assert_equal([0,11], g.links[1].to_coords)
  end

  def test_L_other
    l = "L\t1\t+\t2\t-\t*".to_rgfa_line
    assert_equal(:"2", l.other(:"1"))
    assert_equal(:"1", l.other(:"2"))
    assert_raise(RGFA::NotFoundError){l.other(:"0")}
  end

  def test_L_circular
    l = "L\t1\t+\t2\t-\t*".to_rgfa_line
    assert_equal(false, l.circular?)
    l = "L\t1\t+\t1\t-\t*".to_rgfa_line
    assert_equal(true, l.circular?)
  end

  def test_S
    fields=["S","1","ACGTCACANNN","RC:i:1232","LN:i:11","ab:Z:abcd",
            "FC:i:2321","KC:i:1212"]
    str=fields.join("\t")
    assert_nothing_raised { str.to_rgfa_line }
    assert_equal(RGFA::Line::Segment::GFA1, str.to_rgfa_line.class)
    assert_equal(fields[0].to_sym, str.to_rgfa_line.record_type)
    assert_equal(fields[1].to_sym, str.to_rgfa_line.name)
    assert_equal(fields[2], str.to_rgfa_line.sequence)
    assert_equal(1232, str.to_rgfa_line.RC)
    assert_equal(11, str.to_rgfa_line.LN)
    assert_equal(2321, str.to_rgfa_line.FC)
    assert_equal(1212, str.to_rgfa_line.KC)
    assert_equal("abcd", str.to_rgfa_line.ab)
    assert_raises(RGFA::FormatError) { (str+"\tH1").to_rgfa_line }
    assert_raises(RGFA::FormatError) { "S\tH".to_rgfa_line }
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[2]="!@#?"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::TypeError) do
      f=fields.dup; f[3]="RC:Z:1232"; f.join("\t").to_rgfa_line(version: :gfa1)
    end
    f=["S","2","ACGTCACANNN","LN:i:3"]
    assert_raises(RGFA::InconsistencyError) do
      f.join("\t").to_rgfa_line(vlevel: 1, version: :gfa1)
    end
    f=["S","2","ACGTCACANNN","LN:i:11"]
    assert_nothing_raised { f.join("\t").to_rgfa_line }
    f=["S","2","*","LN:i:3"]
    assert_nothing_raised { f.join("\t").to_rgfa_line }
  end

  def test_forbidden_segment_names
    assert_nothing_raised { "S\tA+B\t*".to_rgfa_line }
    assert_nothing_raised { "S\tA-B\t*".to_rgfa_line }
    assert_nothing_raised { "S\tA,B\t*".to_rgfa_line }
    assert_raises(RGFA::FormatError) do
      "S\tA+,B\t*".to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      "S\tA-,B\t*".to_rgfa_line(vlevel: 1)
    end
  end

  def test_coverage
    l = "S\t0\t*\tRC:i:600\tLN:i:100".to_rgfa_line
    assert_equal(6, l.coverage)
    assert_equal(6, l.coverage!)
    l = "S\t0\t*\tRC:i:600".to_rgfa_line
    assert_equal(nil, l.coverage)
    assert_raises(RGFA::NotFoundError) {l.coverage!}
    l = "S\t0\t*\tLN:i:100".to_rgfa_line
    assert_equal(nil, l.coverage)
    assert_raises(RGFA::NotFoundError) {l.coverage!}
    l = "S\t0\t*\tFC:i:600\tLN:i:100".to_rgfa_line
    assert_equal(nil, l.coverage)
    assert_raises(RGFA::NotFoundError) {l.coverage!}
    assert_equal(6, l.coverage(count_tag: :FC))
    assert_equal(6, l.coverage!(count_tag: :FC))
  end

  def test_P
    fields=["P","4","1+,2-,3+","9M2I3D1M,12M","ab:Z:abcd"]
    str=fields.join("\t")
    assert_nothing_raised { str.to_rgfa_line }
    assert_equal(RGFA::Line::Group::Path, str.to_rgfa_line.class)
    assert_equal(fields[0].to_sym, str.to_rgfa_line.record_type)
    assert_equal(fields[1].to_sym, str.to_rgfa_line.path_name)
    assert_equal([OL[:"1",:"+"],OL[:"2",:"-"],OL[:"3",:"+"]],
                 str.to_rgfa_line.segment_names)
    assert_equal([[RGFA::Alignment::CIGAR::Operation.new(9,:M),
                   RGFA::Alignment::CIGAR::Operation.new(2,:I),
                   RGFA::Alignment::CIGAR::Operation.new(3,:D),
                   RGFA::Alignment::CIGAR::Operation.new(1,:M)],
                  [RGFA::Alignment::CIGAR::Operation.new(12,:M)]],
                 str.to_rgfa_line.overlaps)
    assert_equal("abcd", str.to_rgfa_line.ab)
    assert_raises(RGFA::FormatError) { (str+"\tH1").to_rgfa_line }
    assert_raises(RGFA::FormatError) { "P\tH".to_rgfa_line }
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[2]="1,2,3"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::InconsistencyError) do
      f=fields.dup; f[2]="1+,2+"; f[3]="9M,12M,3M";
                    f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_nothing_raised do
      f=fields.dup; f[3]="*,*"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_nothing_raised do
      f=fields.dup; f[3]="9M2I3D1M,12M,12M"; f.join("\t").
        to_rgfa_line(vlevel: 3)
    end
    assert_nothing_raised do
      f=fields.dup; f[3]="*"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[3]="12,12"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
    assert_raises(RGFA::FormatError) do
      f=fields.dup; f[3]="12M|12M"; f.join("\t").to_rgfa_line(vlevel: 1)
    end
  end

end

File: tests/test_api_gfa1_lines.py
import gfapy
import unittest

class TestApiGfa1Lines(unittest.TestCase):

  def test_C(self):
    fields=["C","1","+","2","-","12","12M","MQ:i:1232","NM:i:3","ab:Z:abcd"]
    s="\t".join(fields)
    gfapy.Line(s) # nothing raised
    self.assertEqual(gfapy.line.edge.Containment, gfapy.Line(s).__class__)
    self.assertEqual(fields[0], gfapy.Line(s).record_type)
    self.assertEqual(fields[1], gfapy.Line(s).from_segment)
    self.assertEqual(fields[2], gfapy.Line(s).from_orient)
    self.assertEqual(fields[3], gfapy.Line(s).to_segment)
    self.assertEqual(fields[4], gfapy.Line(s).to_orient)
    self.assertEqual(12, gfapy.Line(s).pos)
    self.assertEqual([gfapy.CIGAR.Operation(12,"M")], gfapy.Line(s).overlap)
    self.assertEqual(1232, gfapy.Line(s).MQ)
    self.assertEqual(3, gfapy.Line(s).NM)
    self.assertEqual("abcd", gfapy.Line(s).ab)
    with self.assertRaises(gfapy.FormatError): (str+gfapy.Line("\tH1"))
    with self.assertRaises(gfapy.FormatError): gfapy.Line("C\tH")
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[2]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[4]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[5]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[6]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.TypeError):
      f=fields.copy(); f[7]="MQ:Z:1232"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.TypeError):
      f=fields.copy(); f[8]="NM:Z:1232"; gfapy.Line("\t".join(f),vlevel=1)

  def test_L(self):
    fields=["L","1","+","2","-","12M","RC:i:1232","NM:i:3","ab:Z:abcd",
            "FC:i:2321","KC:i:1212","MQ:i:40"]
    s="\t".join(fields)
    gfapy.Line(s) # nothing raised
    self.assertEqual(gfapy.line.edge.Link, gfapy.Line(s).__class__)
    self.assertEqual(fields[0], gfapy.Line(s).record_type)
    self.assertEqual(fields[1], gfapy.Line(s).from_segment)
    self.assertEqual(fields[2], gfapy.Line(s).from_orient)
    self.assertEqual(fields[3], gfapy.Line(s).to_segment)
    self.assertEqual(fields[4], gfapy.Line(s).to_orient)
    self.assertEqual([gfapy.CIGAR.Operation(12,"M")],
                 gfapy.Line(s).overlap)
    self.assertEqual(1232, gfapy.Line(s).RC)
    self.assertEqual(3, gfapy.Line(s).NM)
    self.assertEqual(2321, gfapy.Line(s).FC)
    self.assertEqual(1212, gfapy.Line(s).KC)
    self.assertEqual(40, gfapy.Line(s).MQ)
    self.assertEqual("abcd", gfapy.Line(s).ab)
    with self.assertRaises(gfapy.FormatError): (str+gfapy.Line("\tH1"))
    with self.assertRaises(gfapy.FormatError): gfapy.Line("L\tH")
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[2]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[4]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[5]="x"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.TypeError):
      f=fields.copy(); f[6]="RC:Z:1232"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.TypeError):
      f=fields.copy(); f[7]="NM:Z:1232"; gfapy.Line("\t".join(f),vlevel=1)

  def test_L_coords(self):
    g = gfapy.Gfa(version="gfa1")
    g.append("S\t1\t*\tLN:i:100")
    g.append("L\t1\t+\t2\t-\t1M2D10M1I")
    self.assertEqual(["87","100$"], [str(s) for s in g.dovetails[0].from_coords])
    with self.assertRaises(gfapy.ValueError): g.dovetails[0].to_coords
    g.append("S\t2\t*\tLN:i:100")
    self.assertEqual(["88","100$"], [str(s) for s in g.dovetails[0].to_coords])
    g.append("L\t3\t-\t4\t+\t10M2P3D1M")
    self.assertEqual(["0","14"], [str(s) for s in g.dovetails[1].from_coords])
    self.assertEqual(["0","11"], [str(s) for s in g.dovetails[1].to_coords])

  def test_L_other(self):
    l = gfapy.Line("L\t1\t+\t2\t-\t*")
    self.assertEqual("2", l.other("1"))
    self.assertEqual("1", l.other("2"))
    with self.assertRaises(gfapy.NotFoundError): l.other("0")

  def test_L_circular(self):
    l = gfapy.Line("L\t1\t+\t2\t-\t*")
    self.assertEqual(False, l.is_circular())
    l = gfapy.Line("L\t1\t+\t1\t-\t*")
    self.assertEqual(True, l.is_circular())

  def test_S(self):
    fields=["S","1","ACGTCACANNN","RC:i:1232","LN:i:11","ab:Z:abcd",
            "FC:i:2321","KC:i:1212"]
    s="\t".join(fields)
    gfapy.Line(s) # nothing raised
    self.assertEqual(gfapy.line.segment.GFA1, gfapy.Line(s).__class__)
    self.assertEqual(fields[0], gfapy.Line(s).record_type)
    self.assertEqual(fields[1], gfapy.Line(s).name)
    self.assertEqual(fields[2], gfapy.Line(s).sequence)
    self.assertEqual(1232, gfapy.Line(s).RC)
    self.assertEqual(11, gfapy.Line(s).LN)
    self.assertEqual(2321, gfapy.Line(s).FC)
    self.assertEqual(1212, gfapy.Line(s).KC)
    self.assertEqual("abcd", gfapy.Line(s).ab)
    with self.assertRaises(gfapy.FormatError): (str+gfapy.Line("\tH1"))
    with self.assertRaises(gfapy.FormatError): gfapy.Line("S\tH")
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[2]="!@#?"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.TypeError):
      f=fields.copy(); f[3]="RC:Z:1232"; gfapy.Line("\t".join(f),version="gfa1")
    f=["S","2","ACGTCACANNN","LN:i:3"]
    with self.assertRaises(gfapy.InconsistencyError):
      gfapy.Line("\t".join(f),vlevel=1, version="gfa1")
    f=["S","2","ACGTCACANNN","LN:i:11"]
    gfapy.Line("\t".join(f)) # nothing raised
    f=["S","2","*","LN:i:3"]
    gfapy.Line("\t".join(f)) # nothing raised

  def test_forbidden_segment_names(self):
    gfapy.Line("S\tA+B\t*") # nothing raised
    gfapy.Line("S\tA-B\t*") # nothing raised
    gfapy.Line("S\tA,B\t*") # nothing raised
    with self.assertRaises(gfapy.FormatError):
      gfapy.Line("S\tA+,B\t*",vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      gfapy.Line("S\tA-,B\t*",vlevel=1)

  def test_coverage(self):
    l = gfapy.Line("S\t0\t*\tRC:i:600\tLN:i:100")
    self.assertEqual(6, l.coverage())
    self.assertEqual(6, l.try_get_coverage())
    l = gfapy.Line("S\t0\t*\tRC:i:600")
    self.assertEqual(None, l.coverage())
    with self.assertRaises(gfapy.NotFoundError): l.try_get_coverage()
    l = gfapy.Line("S\t0\t*\tLN:i:100")
    self.assertEqual(None, l.coverage())
    with self.assertRaises(gfapy.NotFoundError): l.try_get_coverage()
    l = gfapy.Line("S\t0\t*\tFC:i:600\tLN:i:100")
    self.assertEqual(None, l.coverage())
    with self.assertRaises(gfapy.NotFoundError): l.try_get_coverage()
    self.assertEqual(6, l.coverage(count_tag="FC"))
    self.assertEqual(6, l.try_get_coverage(count_tag="FC"))

  def test_P(self):
    fields=["P","4","1+,2-,3+","9M2I3D1M,12M","ab:Z:abcd"]
    s="\t".join(fields)
    gfapy.Line(s) # nothing raised
    self.assertEqual(gfapy.line.group.Path, gfapy.Line(s).__class__)
    self.assertEqual(fields[0], gfapy.Line(s).record_type)
    self.assertEqual(fields[1], gfapy.Line(s).path_name)
    self.assertEqual([gfapy.OrientedLine("1","+"),gfapy.OrientedLine("2","-"),
        gfapy.OrientedLine("3","+")],
                 gfapy.Line(s).segment_names)
    self.assertEqual([[gfapy.CIGAR.Operation(9,"M"),
                   gfapy.CIGAR.Operation(2,"I"),
                   gfapy.CIGAR.Operation(3,"D"),
                   gfapy.CIGAR.Operation(1,"M")],
                  [gfapy.CIGAR.Operation(12,"M")]],
                 gfapy.Line(s).overlaps)
    self.assertEqual("abcd", gfapy.Line(s).ab)
    with self.assertRaises(gfapy.FormatError): (str+gfapy.Line("\tH1"))
    with self.assertRaises(gfapy.FormatError): gfapy.Line("P\tH")
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[2]="1,2,3"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.InconsistencyError):
      f=fields.copy(); f[2]="1+,2+"; f[3]="9M,12M,3M"
      gfapy.Line("\t".join(f),vlevel=1)
      f=fields.copy(); f[3]="*,*";
      gfapy.Line("\t".join(f),vlevel=1)
      f=fields.copy(); f[3]="9M2I3D1M,12M,12M";
      gfapy.Line("\t".join(f),vlevel=3)
      f=fields.copy(); f[3]="*";
      gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[3]="12,12"; gfapy.Line("\t".join(f),vlevel=1)
    with self.assertRaises(gfapy.FormatError):
      f=fields.copy(); f[3]="12M|12M"; gfapy.Line("\t".join(f),vlevel=1)

